### PR TITLE
Package dependency resolution

### PR DESCRIPTION
**Pull request by minestarks**
We now resolve dependencies from `qsharp.json`.

Example manifest:

```json
{
  "dependencies": {
    "MyLocalDep": {
      "path": "../some-local-folder"
    },
    "MyRemoteDep": {
      "github": {
        "owner": "some-github-user",
        "repo": "some-github-repo",
        "ref": "commit-or-tag",
      }
    }
  }
}
```

📦 Packages are recursively resolved and their sources added to the current compilation.

📢 Any errors that occur during package resolution is reported through the language service and extension commands.

📝 For remote code, a special text content provider is added so we can show GitHub content as read-only in the editor during debugging and code navigation.

### CODE REVIEW

File: compiler/qsc_project/src/project.rs [hunk 1]
```diff
@@ -164,6 +518,65 @@
...
+pub struct PackageGraphSources {
+    pub root: PackageInfo,
+    pub packages: FxHashMap<PackageKey, PackageInfo>,
+}
+
+impl PackageGraphSources {
+    /// Temporary implementation which just concatenates all sources.
+    /// This will be replaced soon to build and return all the dependencies
+    /// along with the root package sources.
+    /// TODO(packages): delete this method and use the package graph instead
```
[the remark this answers is not in the document]
Comment (minestarks) [reply compiler/qsc_project/src/project.rs:568]: Known, will address in next PR

File: language_service/src/state.rs [hunk 2]
```diff
@@ -262,28 +258,13 @@
 
         let removed_compilation = self.remove_open_document(uri);
 
+        // TODO: consume dependency sources as well
```
[the remark this answers is not in the document]
Comment (minestarks) [reply language_service/src/state.rs:261]: Known, will address in next PR

File: pip/tests/test_project.py [hunk 3]
```diff
@@ -75,6 +87,34 @@
             },
             "qsharp.json": "{}",
         },
+        "with_deps": {
+            "src": {
+                # TODO(packages) When package aliases are taken into account,
```
[the remark this answers is not in the document]
Comment (minestarks) [reply pip/tests/test_project.py:92]: Known, will address in next PR

File: npm/qsharp/src/browser.ts [hunk 4]
```diff
@@ -172,7 +173,22 @@ export type {
...
+  type Example,
+  type Exercise,
+  type ExplainedSolution,
+  type ExplainedSolutionItem,
+  type Kata,
+  type KataSection,
+  type Lesson,
+  type LessonItem,
+  type Question,
+} from "./katas.js";
```
Comment (swernli) [inline npm/qsharp/src/browser.ts:190]: This seems unrelated, could it be from a merge artifact?

File: npm/qsharp/src/language-service/language-service.ts [hunk 5]
```diff
@@ -117,6 +118,7 @@ export class QSharpLanguageService implements ILanguageService {
       readFile: async () => null,
       listDirectory: async () => [],
       resolvePath: async () => null,
+      fetchGithub: async () => "",
```
Comment (swernli) [inline npm/qsharp/src/language-service/language-service.ts:121]: Note from Mine's past self: would it make sense to provide a default implementation that the playground can also use?

File: npm/qsharp/src/compiler/compiler.ts [hunk 6]
```diff
@@ -144,7 +153,6 @@ export class Compiler implements ICompiler {
       profile ?? "adaptive_ri",
     );
   }
```
Comment (swernli) [inline npm/qsharp/src/compiler/compiler.ts:155]: nitpick: newline is gone?

File: npm/qsharp/src/compiler/compiler.ts [hunk 7]
```diff
@@ -274,3 +296,18 @@ export const compilerProtocol: ServiceProtocol<ICompiler, QscEventData> = {
...
+    ? {
+        root: {
+          sources: program.sources,
+          languageFeatures: program.languageFeatures || [],
+          dependencies: {},
+        },
+        packages: {},
+      }
+    : program.packageGraphSources;
+}
```
Comment (swernli) [inline npm/qsharp/src/compiler/compiler.ts:313]: This appears to be dead code.

```suggestion
```

File: pip/tests/test_project.py [hunk 8]
```diff
@@ -51,11 +51,23 @@ def test_project_unreadable_source(qsharp) -> None:
...
+    assert result == 4
+
+
+def test_project_circular_dependency_error(qsharp) -> None:
+    with pytest.raises(Exception) as excinfo:
+        qsharp.init(project_root="/circular")
+    assert str(excinfo.value).find("Circular dependency detected between") != -1
+
+
 memfs = {
```
Comment (swernli) [inline pip/tests/test_project.py:66]: Would it make sense to have some fake github sources here to test that functionality in Python too?

File: vscode/qsharp.schema.json [hunk 9]
```diff
@@ -33,6 +33,9 @@
           }
         }
       }
+    },
+    "dependencies": {
```
Comment (swernli) [inline vscode/qsharp.schema.json:37]: There might be more fields that are needed here.

File: vscode/src/projectSystem.ts [hunk 10]
```diff
@@ -132,14 +133,16 @@ async function readFileUri(
...
     // `readFile` should throw the below if the file is not found
     if (
-      !(_err instanceof vscode.FileSystemError && _err.code === "FileNotFound")
+      !(err instanceof vscode.FileSystemError && err.code === "FileNotFound")
     ) {
-      log.error("Unexpected error trying to read file", _err);
+      log.error("Unexpected error trying to read file", err);
+      return null;
+    } else {
+      throw err;
```
Comment (swernli) [inline vscode/src/projectSystem.ts:144]: Maybe it's ok to always throw here? Makes it less likely errors will be missed and require looking at the logs.

File: vscode/test/suites/language-service/language-service.test.ts [hunk 11]
```diff
@@ -174,4 +202,79 @@ suite("Q# Language Service Tests", function suite() {
...
+      vscode.languages.getDiagnostics(projectMainDocUri);
+
+    assert.isEmpty(
+      actualDiagnostics,
+      `Expected no diagnostics, but got ${JSON.stringify(actualDiagnostics)}`,
+    );
+
+    // Sanity check the test setup - is this the correct position?
+    const text = doc.getText(
+      new vscode.Range(new vscode.Position(3, 16), new vscode.Position(3, 26)),
```
Comment (swernli) [inline vscode/test/suites/language-service/language-service.test.ts:221]: Looks like these offsets need to get updated to match the below.

File: language_service/src/compilation.rs [hunk 12]
```diff
@@ -90,8 +91,9 @@ impl Compilation {
         Self {
             package_store,
             user_package_id: package_id,
-            errors,
+            compile_errors: errors,
```
Comment (swernli) [inline language_service/src/compilation.rs:94]: nitpick (from @ScottCarda-MS): rename `errors` to `compile_errors`

File: language_service/src/tests.rs [hunk 13]
```diff
@@ -185,6 +188,7 @@ async fn document_in_project() {
                             ),
                         ),
                     ],
+                    [],
```
Comment (swernli) [inline language_service/src/tests.rs:191]: Is it worth adding a test case that propagates up a project error?

File: wasm/src/project_system.rs [hunk 14]
```diff
@@ -141,36 +188,110 @@ impl ProjectLoader {
...
+        project_config.try_into()
     }
 }
 
-impl From<ProjectSources> for Vec<(String, String)> {
-    fn from(sources: ProjectSources) -> Self {
-        serde_wasm_bindgen::from_value(sources.into())
-            .expect("sources object should be an array of string pairs")
-    }
+#[allow(clippy::needless_pass_by_value)]
```
Comment (swernli) [inline wasm/src/project_system.rs:208]: It looks like this can be removed

**Review by swernli — APPROVED** [review]

File: npm/qsharp/src/language-service/language-service.ts (hunk 5 above)
Comment (minestarks) [reply npm/qsharp/src/language-service/language-service.ts:121]: Reply from my current self: Yes, that would be useful, but I don't have the time right now :)

File: npm/qsharp/src/compiler/compiler.ts [hunk 15]
```diff
@@ -74,11 +75,18 @@ export interface ICompiler {
  * Type definition for the configuration of a program.
  * If adding new properties, make them optional to maintain backward compatibility.
  */
-export type ProgramConfig = {
-  /** An array of source objects, each containing a name and contents. */
-  sources: [string, string][];
-  /** An array of language features to be opted in to in this compilation. */
-  languageFeatures?: string[];
+export type ProgramConfig = (
```
[the remark this answers is not in the document]
Comment (minestarks) [reply npm/qsharp/src/compiler/compiler.ts:78]: Well, I'd describe it a bit more differently, and I'd say the type _is_ fulfilling its job. It:

- Encapsulates all the program-related parameters so we don't have to add backcompat overloads or _every single function_ that takes a program. We just need to update this type
- Gives us the flexibility to reorder parameters by making everything named (e.g. so we don't end up with code like `compile(undefined, undefined, someparam)` in QCOM

Also, I don't really consider single-file mode "legacy" - for example, the Playground will be single file mode for the foreseeable future. I think there's value in allowing a simpler interface where you can pass in sources directly like this.

I honestly think this type is fine, and it might be TypeScript itself that's bothering you? It's an input object that can either have `sources` or `packageGraphSources` and I think that's just fine.
